engines/immortal/kernal.cpp

```diff
@@ -0,0 +1,1107 @@
+	count = dir * 256;
+
+	while ((count >= 0) && (count <= 256)) {
+		fadePal(pal, count, target);
+		Utilities::delay8(delay);
+		setColors(target);
+
+		// Same as above, it was originally a branch, this does the same thing
+		count += (dir == 0) ? 16 : -16;
+	}
+}
+
+// These two can probably be removed and instead use an enum to declare fadeout/in
+void ImmortalEngine::fadeOut(int j) {
+	fade(_palDefault, 1, j);
+}
+
+void ImmortalEngine::normalFadeOut() {
+	fadeOut(15);
+}
+
+void ImmortalEngine::slowFadeOut() {
+	fadeOut(28);
+}
+
+void ImmortalEngine::fadeIn(int j) {
+	fade(_palDefault, 0, j);
+}
+
+void ImmortalEngine::normalFadeIn() {
+	fadeIn(15);
+}
+
+// These two can probably be removed since the extra call in C doesn't have the setup needed in ASM
+void ImmortalEngine::useBlack() {
+	setColors(_palBlack);
+}
+void ImmortalEngine::useWhite() {
+	setColors(_palBlack);
+}
+
+void ImmortalEngine::useNormal() {
+	setColors(_palDefault);
+	 _usingNormal = 1;
+}
+
+void ImmortalEngine::useDim() {
+	setColors(_palDim);
+	_usingNormal = 0;
+}
+
+
+/*
+ *
+ * -----                 -----
+ * ----- Input Functions -----
+ * -----                 -----
+ *
+ */
+
+void ImmortalEngine::waitKey() {
+	bool wait = true;
+	while (wait == true) {
+		if (getInput() == true) {
+			wait = false;
+		}
+	}
+}
+
+// This was originally in Motives, which is weird since it seems more like an engine level function, so it's in kernal now
+void ImmortalEngine::waitClick() {
+	bool wait = true;
+	while (wait == true) {
+		if (getInput() == true) {
+			Utilities::delay(25);
+			if (buttonPressed() || firePressed()) {
+				wait = false;
+			}
+		}
+	}
+}
+
+// These functions are not yet implemented
+void ImmortalEngine::blit8() {}
+void ImmortalEngine::addKeyBuffer() {}
+void ImmortalEngine::clearKeyBuff() {}
+void ImmortalEngine::userIO() {}
+void ImmortalEngine::pollKeys() {}
+void ImmortalEngine::noNetwork() {}
+bool ImmortalEngine::getInput() {
+	return true;
+}
+// ----
+
+/*
+ *
+ * -----                       -----
+ * ----- Sound/Music Functions -----
+ * -----                       -----
+ *
+ */
+
+void ImmortalEngine::toggleSound() {
+	// Interestingly, this does not mute or turn off the sound, it actually pauses it
+	_themePaused = !_themePaused;
+	fixPause();
+}
+
+void ImmortalEngine::fixPause() {
+	/* The code for this is a little strange, but the idea is that you have
+	 * a level theme, and a combat theme, that can both be active. So first you
+	 * pause the level theme, and then you pause the combat theme.
+	 * The way it does it is weird though. Here's the logic:
+	 * if playing either text or maze song, check if the theme is paused. else, just go ahead and pause.
+	 * Same thing for combat song. A little odd.
+	 */
+
+	// This is a nasty bit of code isn't it? It's accurate to the source though :D
+	switch (_playing) {
+		case kSongText:
+		case kSongMaze:
+			if (_themePaused) {
+				musicUnPause(_themeID);
+				break;
+			}
+		default:
```

Comment:
If the break is intentionally missing here, please add a comment `// fall through`